src/app/destinyTrackerApi/d2-itemListBuilder.ts

```diff
@@ -8,83 +7,73 @@ import { D2ItemFetchRequest } from '../item-review/d2-dtr-api-types';
 import { translateToDtrItem } from './d2-itemTransformer';
 
 /**
- * Translates collections of DIM items into a collection of data almost ready to ship to the DTR API.
- * Generally tailored to work with weapon data.
- * Also tailored for the D2 version.
+ * Translate the universe of items that the user has in their stores into a collection of data that we can send the DTR API.
+ * Tailored to work alongside the bulkFetcher.
+ * Non-obvious bit: it attempts to optimize away from sending items that already exist in the ReviewDataCache.
  */
-class D2ItemListBuilder {
-  _getNewItems(allItems: D2Item[], reviewDataCache: D2ReviewDataCache) {
-    const allDtrItems = allItems.map(translateToDtrItem);
-    const allKnownDtrItems = reviewDataCache.getItemStores();
+export function getItemList(stores: D2Store[], reviewDataCache: D2ReviewDataCache): D2ItemFetchRequest[] {
+  const dtrItems = getDtrItems(stores, reviewDataCache);
 
-    const unmatched = allDtrItems.filter((di) => allKnownDtrItems.every((kdi) => di.referenceId !== kdi.referenceId));
+  const list = new Set(dtrItems);
 
-    return unmatched;
-  }
-
-  _getNewVendorItems(vendorItems: D2ItemFetchRequest[], reviewDataCache: D2ReviewDataCache) {
-    const allKnownDtrItems = reviewDataCache.getItemStores();
-
-    const unmatched = vendorItems.filter((vi) => allKnownDtrItems.every((di) => di.referenceId !== vi.referenceId));
-
-    return unmatched;
-  }
+  return Array.from(list);
+}
 
-  /**
-   * In D1, extracting all of the items from all of the stores was a little different from
-   * extracting all of the vendor items. If the interface is the same for both, this can go
-   * away.
-   */
-  _getAllItems(stores: D2Store[]): D2Item[] {
-    return flatMap(stores, (store) => store.items);
+export function getVendorItemList(
+  reviewDataCache: D2ReviewDataCache,
+  vendorSaleItems?: DestinyVendorSaleItemComponent[],
+  vendorItems?: DestinyVendorItemDefinition[]
+): D2ItemFetchRequest[] {
+  if (vendorSaleItems) {
+    const allVendorItems = vendorSaleItems.map((vendorItem): D2ItemFetchRequest => ({ referenceId: vendorItem.itemHash }));
+
+    return getNewVendorItems(allVendorItems, reviewDataCache);
+  } else if (vendorItems) {
+    const allVendorItems = vendorItems.map((vi) => ({ referenceId: vi.itemHash })) as D2ItemFetchRequest[];
+
+    return getNewVendorItems(allVendorItems, reviewDataCache);
+  } else {
+    throw new Error("Neither sale items nor vendor items were supplied.");
   }
+}
 
-  // Get all of the weapons from our stores in a DTR API-friendly format.
-  _getDtrItems(stores: D2Store[], reviewDataCache: D2ReviewDataCache): D2ItemFetchRequest[] {
-    const allItems = this._getAllItems(stores);
+function getNewItems(allItems: D2Item[], reviewDataCache: D2ReviewDataCache) {
```

Comment:
Did the functions get put in alphabetical order?

The diff is making me dizzy.

Reply:
I moved the exported functions to the top, then outdented them all since they aren’t in a class. GitHub loves whitespace-sensitive diffs.